### ISSUE TITLE
Update to support new features in wcslib 7.1

### ISSUE THREAD

**Issue by Cadair** [sunpy/sunpy#3643]
wcslib 7.1 includes:

>  Support Solar keywords RSUN_REF, DSUN_OBS, CRLN_OBS, CRLT_OBS, HGLN_OBS, and HGLT_OBS by accomodating them within an auxiliary struct, auxprm, within the wcsprm struct.  Now filled by the header parsers, wcspih() and wcsbth() and handled routinely by wcsinit(), wcssub(), wcscompare(), wcsfree(), wcsprt(), wcsset(), and wcshdo().
Requested by Stuart Mumford (SunPy) with input from Bill Thompson.

While we can't rely on wcslib 7.1 yet, as it's not even been merged into astropy (astropy/astropy#9829), we can work out how to support it in parallel with our old hacks.

What this allows us to do is to use the underlying wcslib strcuts for storing the heliographic observer location and rsun reference value. The most important place to support the new functionality is in the code which converts WCSes to and from coordinate frames [here](https://github.com/sunpy/sunpy/blob/master/sunpy/coordinates/wcs_utils.py#L22-L30) and [here](https://github.com/sunpy/sunpy/blob/master/sunpy/coordinates/wcs_utils.py#L56-L64) as this will support people loading compatible headers directly with `astropy.wcs.WCS`.

We also then need to update [`GenericMap.wcs`](https://github.com/sunpy/sunpy/blob/master/sunpy/map/mapbase.py#L265) to "do the right thing" with ensuring that the information is in the struct, and more importantly is consistent with the `GenericMap.observer_coordinate` property, which means removing any inconsistent HG or CR headers and converting the `observer_coordinate` back to HG and putting back in the header at the right place.

Once all of this is done, we need to make sure we test all these options, as well as adding tests for direct parsing of `WCS` objects and headers to `sunpy.Map`, and through the WCS <> coordinate converters. Then finally we need to document this madness somewhere.  

Addressing this would close #2910

**Comment by Cadair** [comment]
It should be noted that 7.1 is now in astropy master and will be included in 4.0.1, so this is now easy to hack on.

**Comment by dstansby** [comment]
To add an example, the goal is to make this code add the observer coordinate properly:

```python
from astropy.wcs.utils import wcs_to_celestial_frame
from astropy.wcs import WCS
from sunpy.util import MetaDict
import sunpy.coordinates.wcs_utils

meta = MetaDict([('wcsaxes', 2), ('crpix1', 512.5), ('crpix2', 512.5),
                 ('cdelt1', 1.953125), ('cdelt2', 1.953125),
                 ('cunit1', 'arcsec'), ('cunit2', 'arcsec'),
                 ('ctype1', 'HPLN-TAN'), ('ctype2', 'HPLT-TAN'),
                 ('crval1', 0.0), ('crval2', 0.0),
                 ('lonpole', 180.0), ('latpole', 0.0),
                 ('date-obs', '2020-01-01T00:00:00.000'),
                 ('hgln_obs', 0.0), ('hglt_obs', -2.942433009185766),
                 ('dsun_obs', 147098545612.676),
                 ('rsun_ref', 695700000.0), ('rsun_obs', 975.5185219424909)])
wcs = WCS(meta)
print(wcs.heliographic_observer)

frame = wcs_to_celestial_frame(wcs)
print(frame)
```

Note how the observer is currently `None`:
```
WCS Keywords

Number of WCS axes: 2
CTYPE : 'HPLN-TAN'  'HPLT-TAN'  
CRVAL : 0.0  0.0  
CRPIX : 512.5  512.5  
PC1_1 PC1_2  : 1.0  0.0  
PC2_1 PC2_2  : 0.0  1.0  
CDELT : 0.0005425347222222222  0.0005425347222222222  
NAXIS : 0  0
<Helioprojective Frame (obstime=2020-01-01T00:00:00.000, rsun=695700.0 km, observer=None)>
```

**Comment by Cadair** [comment]
@DanRyanIrish and @hayesla have been working on constructing WCSes where the observer is specified by an astropy `EarthLocation` object, which is then converted to `SkyCoord` and attached to the WCS.

The changes to wcslib which let it parse the `hgln/lt` headers are great but fail to account for this use case. Maybe we would want a FITS WCS subclass in SunPy which provides some extra functionality for setting observer location in the Python API (i.e. `observer_coordinate`) which then generates the `HGLN/LT` headers in the wcslib struct.

**Comment by astrofrog** [comment]
For observers that are earth locations, is there any reason not to use ``OBSGEO-?`` keywords?

**Comment by Cadair** [comment]
I was quite surprised to see when talking to @hayesla that astropy's `wcs_to_celestial_frame` didn't use `obsgeo-` I don't think either me or @hayesla opened an astropy issue for that.
